larray/tests/test_constrained_session.py

```diff
@@ -0,0 +1,704 @@
+
+
+@needs_pytables
+def test_init_constrainedsession_hdf():
+    cs = TestConstrainedSession(inputpath('test_session.h5'))
+    assert set(cs.keys()) == {'b', 'b024', 'a', 'a2', 'anonymous', 'a01', 'ano01', 'c', 'd', 'e', 'g', 'f', 'h'}
+
+
+def test_getitem_cs(constrainedsession):
+    test_getitem(constrainedsession)
+
+
+def test_setitem_cs(constrainedsession):
+    cs = constrainedsession
+
+    # only change values of an array -> OK
+    cs['h'] = zeros_like(h)
+
+    # trying to add an undeclared variable -> prints a warning message
+    with pytest.warns(UserWarning) as caught_warnings:
+        cs['i'] = ndtest((3, 3))
+    assert len(caught_warnings) == 1
+    assert caught_warnings[0].message.args[0] == f"'i' is not declared in '{cs.__class__.__name__}'"
+
+    # trying to set a variable with an object of different type -> should fail
+    # a) type given explicitly
+    # -> Axis
+    expected_error_msg = "instance of Axis expected"
+    with pytest.raises(TypeError) as error:
+        cs['a'] = 0
+    assert str(error.value) == expected_error_msg
+    # -> ConstrainedArray
+    expected_error_msg = "Expected object of type 'Array' or a scalar for the variable 'h' " \
+                         "but got object of type 'ndarray'"
+    with pytest.raises(TypeError) as error:
+        cs['h'] = h.data
+    assert str(error.value) == expected_error_msg
+    # b) type deduced from the given default value
+    expected_error_msg = "instance of Axis expected"
+    with pytest.raises(TypeError) as error:
+        cs['b'] = ndtest((3, 3))
+    assert str(error.value) == expected_error_msg
+
+    # trying to set a ConstrainedArray variable using a scalar -> OK
+    cs['h'] = 5
+
+    # trying to set a ConstrainedArray variable using an array with axes in different order -> OK
+    cs['h'] = h.transpose()
+
+    # trying to set a ConstrainedArray variable using an array with wrong axes -> should fail
+    # a) missing axis
+    expected_error_msg = "Array 'h' was declared with axes {a, b} but got array with axes {a} ({b} axis is missing)"
+    with pytest.raises(ValueError) as error:
+        cs['h'] = ndtest(a3)
+    assert str(error.value) == expected_error_msg
+    # b) extra axis
+    expected_error_msg = "Array 'h' was declared with axes {a, b} but got array with axes {a, b, c} " \
+                         "(unexpected {c} axis)"
+    with pytest.raises(ValueError) as error:
+        cs['h'] = ndtest((a3, b2, 'c=c0..c2'))
+    assert str(error.value) == expected_error_msg
+    # c) incompatible axis
+    expected_error_msg = """\
+Incompatible axis for array 'h':
+Axis(['a0', 'a1', 'a2', 'a3', 'a4'], 'a')
+vs
+Axis(['a0', 'a1', 'a2', 'a3'], 'a')"""
+    with pytest.raises(ValueError) as error:
+        cs['h'] = h.append('a', 0, 'a4')
+    assert str(error.value) == expected_error_msg
+
+    # trying to set a ConstrainedArray variable using an array with wrong dtype -> print a warning
+    with pytest.warns(UserWarning) as caught_warnings:
```

Comment:
I removed the warning message in case of wrong dtype in commit cbff52c6c844398485114f2969bf21d706db31dc (too much annoying)

Reply:
I forgot to comment about this... I think the best situation would be to check if the value dtype can be safely casted to the "destination" dtype and if so, just do it (no warning at all) but if it's not the case (e.g. when storing float values in an integer array) either warn about it or even raise an exception.

Reply:
np.can_cast should be useful for this, I think